### PR TITLE
Vectorize UTF16 offset calculations

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
Benchmark numbers look pretty good 🙂

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

File: stdlib/public/core/StringUTF16View.swift [hunk 1]
```diff
@@ -532,9 +532,12 @@ extension String.UTF16View {
     let idx = _utf16AlignNativeIndex(idx)
 
     guard _guts._useBreadcrumbs(forEncodedOffset: idx._encodedOffset) else {
-      // TODO: Generic _distance is still very slow. We should be able to
-      // skip over ASCII substrings quickly
-      return _distance(from: startIndex, to: idx)
+      return _guts.withFastUTF8 { utf8 in
+        let sliced = UnsafeBufferPointer(
+          rebasing: utf8[startIndex._encodedOffset ..< idx._encodedOffset]
```
Comment (lorentey) [inline stdlib/public/core/StringUTF16View.swift:537]: This assumes that `idx` has already been validated to be within bounds, which isn't currently the case. 

(I'm planning to systematically address this in https://github.com/apple/swift/pull/41417, but I haven't got to the UTF-16 view yet!)

File: stdlib/public/core/UnicodeHelpers.swift [hunk 2]
```diff
@@ -122,6 +122,46 @@ internal func _utf8ScalarLength(_ x: UInt8) -> Int {
   return (~x).leadingZeroBitCount
 }
 
+@inlinable @inline(__always)
+internal func _utf16Length(_ rawBuffer: UnsafeRawBufferPointer) -> Int {
+  guard rawBuffer.count > 0 else { return 0 }
+  _internalInvariant(!UTF8.isContinuation(rawBuffer.first.unsafelyUnwrapped))
+  _internalInvariant(
+       UTF8.isASCII(rawBuffer.last.unsafelyUnwrapped)
+    || UTF8.isContinuation(rawBuffer.last.unsafelyUnwrapped))
```
Comment (lorentey) [inline stdlib/public/core/UnicodeHelpers.swift:142]: Similarly, `idx` may not be scalar-aligned, in which case the slice may end up starting/ending on partial scalars, violating these invariants.

File: stdlib/public/core/UnicodeHelpers.swift [hunk 3]
```diff
@@ -122,6 +122,46 @@ internal func _utf8ScalarLength(_ x: UInt8) -> Int {
...
+  let wordASCIIMask = UInt(truncatingIfNeeded: 0x8080_8080_8080_8080 as UInt64)
+  var utf16Count = 0
+  var readIdx = 0
+
+  while readIdx + MemoryLayout<UInt>.stride < rawBuffer.count {
+    let maybeASCII = Builtin.loadRaw((rawBuffer.baseAddress._unsafelyUnwrappedUnchecked + readIdx)._rawValue) as UInt
```
Comment (lorentey) [inline stdlib/public/core/UnicodeHelpers.swift:148]: Is this how to spell an unaligned load? I think we should rather put this in a utility function that is named to indicate this.

(Would arranging for aligned loads be better? IIRC, Hasher's buffer combining method aligns its loads, but I don't recall if I ever benchmarked it.)

I wonder if there is some neat simd-style trick to count bytes that don't have `10` in their most significant bits. E.g., `((word & 0xC0C0C0C0...) ^ 0x80808080...)` reduces it to counting non-zero bytes. Hm, this could be promising... 🤔

Comment (Catfish-Man) [reply stdlib/public/core/UnicodeHelpers.swift:148]: Unaligned loads are going through S-E shortly :D I'm just gonna wait for that. And yeah I would love to find some good tricks here.

I didn't do aligned loads because it's easy to fall back into unaligned if the non-ascii cleanup loop ends in a multibyte character, but it would be possible to fix that up if it turned out to be profitable.

File: stdlib/public/core/UnicodeHelpers.swift (hunk 2 above)
Comment (Catfish-Man) [reply stdlib/public/core/UnicodeHelpers.swift:142]: I think in practice it always is for this caller, but I wasn't 100% sure, so I put the invariants in to catch if I was wrong :)

File: stdlib/public/core/StringUTF16View.swift (hunk 1 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringUTF16View.swift:537]: eep! Thank you

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci Apple Silicon benchmark

File: validation-test/stdlib/StringBreadcrumbs.swift [hunk 4]
```diff
@@ -1,5 +1,3 @@
-// rdar://84233775
-// REQUIRES: rdar84233775
```
Comment (Catfish-Man) [inline validation-test/stdlib/StringBreadcrumbs.swift:2]: This is passing for me now, but I would be fine with separating out reenabling it

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please test

File: stdlib/public/core/StringUTF16View.swift [hunk 5]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
   }
 }
 
+// FIXME: This is for debugging only; remove before merging.
+extension String.Index: CustomStringConvertible {
+  @_alwaysEmitIntoClient
+  @inline(never)
+  public var description: String {
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift:525]: Oops, this was intended to be temporary, will remove

File: stdlib/public/core/StringUTF16View.swift [hunk 6]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
...
+    var utf16Count = 0
+    
+    while readPtr + MemoryLayout<U>.stride < endPtr {
+      //Find the number of continuations (0b10xxxxxx)
+      let sValue = readPtr.load(as: S.self)
+      let continuations = S.zero.replacing(with: S.one, where: sValue .< -65 + 1)
+      let continuationCount = Int(continuations.wrappedSum())
+            
+      //Find the number of 4 byte code points (0b11110xxx)
+      let uValue = readPtr.load(as: U.self)
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift:570]: I'm not delighted that I have to reload it like this, suggestions on avoiding it would be appreciated

File: stdlib/public/core/StringUTF16View.swift [hunk 7]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
...
+    endPtr: UnsafeRawPointer,
+    unsignedSIMDType: U.Type,
+    signedSIMDType: S.Type
+  ) -> Int where U.Scalar == UInt8, S.Scalar == Int8 {
+    var utf16Count = 0
+    
+    while readPtr + MemoryLayout<U>.stride < endPtr {
+      //Find the number of continuations (0b10xxxxxx)
+      let sValue = readPtr.load(as: S.self)
+      let continuations = S.zero.replacing(with: S.one, where: sValue .< -65 + 1)
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift:566]: -65 + 1 is from Daniel Lemire's version of this. I left it alone for now, but it definitely reads oddly.

File: stdlib/public/core/StringUTF16View.swift [hunk 8]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
...
+        let len = _utf8ScalarLength(byte)
+        // if we don't have enough bytes left, we don't have a complete scalar,
+        // so don't add it to the count.
+        if readPtr + len <= endPtr {
+          utf16Count &+= len == 4 ? 2 : 1
+        }
+        readPtr += len
+      }
+      
+    //  utf16Count &+= _utf16Length(
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift:618]: Wider vector sizes are much slower in my tests. Steve mentioned there's an underlying LLVM change that needs to be adjusted for in the SIMD types.

File: stdlib/public/core/StringUTF16View.swift [hunk 9]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
...
+
+      // Currently, using SIMD sizes above SIMD8 is slower
+      utf16Count &+= _utf16Length(
+        readPtr: &readPtr,
+        endPtr: endPtr,
+        unsignedSIMDType: SIMD8<UInt8>.self,
+        signedSIMDType: SIMD8<Int8>.self
+      )
+    //
+    //  utf16Count &+= _utf16Length(
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift:647]: Do we want smaller vectors for tail cleanup or just go straight to byte by byte?

File: stdlib/public/core/StringUTF16View.swift [hunk 10]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
...
+      readPtr += MemoryLayout<U>.stride
+    }
+    
+    return utf16Count
+  }
+  
+  @inline(__always)
+  internal func _utf16Distance(from start: Index, to end: Index) -> Int {
+    _internalInvariant(end.transcodedOffset == 0 || end.transcodedOffset == 1)
+    return (end.transcodedOffset - start.transcodedOffset) + _guts.withFastUTF8(
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift:585]: This handling of transcoded offsets makes some sense to me and passes all tests, but I'd appreciate eyes on it, since it's the bit I'm least confident in

**Comment by Catfish-Man** [comment]
I'm probably going to make a clean PR for this once it's ready, since this had a few experiments in it before I settled on what approach I was taking

File: stdlib/public/core/StringUTF16View.swift [hunk 11]
```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
...
+      while readPtr < endPtr {
+        let byte = readPtr.load(as: UInt8.self)
+        if !UTF8.isContinuation(byte) {
+          break
+        }
+        readPtr += 1
+      }
+      
+     // align
+      while readPtr < endPtr
```
Comment (stephentyrone) [inline stdlib/public/core/StringUTF16View.swift:606]: We should really just write this operation in asm for x86_64 and arm64 so that we don't have to do this dance.

I'm happy to do that in a separate PR though. =)

Comment (stephentyrone) [reply stdlib/public/core/StringUTF16View.swift:606]: Actually, for the Swift version you should not bother aligning your reads. Just start from wherever you start; given the usual alignment and length distributions these are going to see, the cost of unaligned loads is less than the cost of initial scalar work to get alignment.

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Cleaned up PR here https://github.com/apple/swift/pull/41866, will close this once tests and benchmarks finish running
